Maintain New Invariant For synchronous_mode_strict

Maithem opened the pull request:
This patch adds a new DCS update rule to maintain at least one synchronous replica in /sync. Prevent updates with an empty set by keeping the last known synchronous replica. This increases the chances of being able to elect a new leader. For example, if there are three nodes A, B and C (where A is the leader) then the following sequence prevents electing a new leader:

1. Shutdown B
2. Shutdown C
3. Shutdown A
4. Start B
5. Start C

If /sync is set '*' after step #2 then a new leader cannot be elected when B and C are up despite one of them being a synchronous replica. This patch skips updating the DCS such that sync_standby will still contain C, which will allow the cluster to recover correctly by promoting node C if A never joins back.

Addresses #3327

CyberDem0n commented on patroni/ha.py:
```diff
@@ -914,6 +914,13 @@ def _process_multisync_replication(self) -> None:
+
+        # Since synchronous strict mode requires at least one other replica, /sync should not be updated
+        # with an empty set
+        if global_config.is_synchronous_mode_strict and len(voters) != 0 and not sync_common:
```
What about the following case?
We have a primary A and sync replica B.
At the same time when B is gone C is connected.
It will lead to the situation that `synchronous_standby_names` and `/sync` key will not be updated.
I don't think it is desired behavior.

Maithem replied:
Hi @CyberDem0n, hit this issue again, so coming back to it.

Your "B gone while C connects" case is covered. When picked is non-empty (a streaming replacement exists), we skip the early return, and the final write_sync_state call in _process_multisync_replication sets both synchronous_standby_names and /sync to {C} in one cycle. The new if not picked: return fires only when picked is also empty. That's the original #3327 reproducer.

Added a unit test for that scenario: tests/test_ha.py::TestHa::test__process_multisync_replication, the "/sync is updated (not blocked) when a new replica C replaces departed replica B" sub-case. Plus a non-strict-mode regression test and a multi-voter narrowing test. All green.

#3435 is a broader strict-mode rewrite. Could we land this as the targeted fix and treat #3435 as the larger refactor on its own track? I can also backport this to the 4.x patch line if that helps.

CyberDem0n replied:
@Maithem it would be very helpful if you spend some time testing #3435.
However, I didn't plan to backport it to older major/minor releases because it is not a simple bugfux, but a big architectural change.

Maithem replied:
@CyberDem0n we're on 4.0.x today and planning to move to the 4.x release carrying #3435 once it lands — happy to test #3435 along the way. In parallel, can we push this PR through? Without an upstream backport we'd have to carry it as an internal fork, which we'd really like to avoid. I'll own the rebase onto `REL_4_0` (and `REL_4_1`) and keep them in sync as #3435 evolves.

Maithem replied:
@CyberDem0n I need to know how to proceed, please make a call ^^

CyberDem0n replied:
This PR solves the same problem as #3435, however this solution is incomplete and doesn't cover many edge cases. Why should it be pushed?

Maithem replied:
- Closes the referenced bug (#3327) with a small diff behind a strict-mode only guard
- Backportable to REL_4_0
- Scope is synchronous_mode: true only; quorum is intentionally left to #3435

If there's something specific inside the synchronous_mode: true window you want covered before this lands, happy to extend, otherwise asking to merge as scoped.

Maithem committed:
Maintain New Invariant For synchronous_mode_strict

This patch adds new update rules to maintain the last known
synchronous replicas in both /sync and synchronous_standby_names
when strict mode is enabled.  Prevent updates that would clear
/sync or downgrade synchronous_standby_names to '*' when the last
sync replica disappears with no streaming replacement.  This
increases the chances of being able to elect a new leader and
ensures commits remain pinned to the actual lost member(s) rather
than being satisfied by any reconnecting async replica.

For example, if there are three nodes A, B and C, where A is the
leader and synchronous_node_count is 2, then the following sequence
prevents electing a new leader:

1. Shutdown B
2. Shutdown C
3. Shutdown A
4. Start B
5. Start C

Without this patch, after step #2 /sync is cleared and
synchronous_standby_names is set to '*', so a new leader cannot be
elected when B and C come back despite one of them being a sync
replica moments ago.  With this patch /sync still contains C and
synchronous_standby_names still names C, which allows the cluster
to recover correctly by promoting node C if A never joins back,
and blocks unsafe commits in the meantime.

hughcapet commented:
resolved via https://github.com/patroni/patroni/pull/3435
